File: typescript/tests/first.test.ts
// ===========================================================================
// SECTION 8: FIRST (ORDERED CHOICE) (8 tests)
// ===========================================================================

import { testParse } from './testUtils.js';

describe('First (Ordered Choice)', () => {
  test('FR01-match 1st', () => {
    const { ok, errorCount } = testParse(
      'S <- "abc" / "ab" / "a" ;',
      'abc'
    );
    expect(ok).toBe(true);
    expect(errorCount).toBe(0);
  });

  test('FR02-match 2nd', () => {
    const { ok, errorCount } = testParse(
      'S <- "xyz" / "abc" ;',
      'abc'
    );
    expect(ok).toBe(true);
    expect(errorCount).toBe(0);
  });

  test('FR03-match 3rd', () => {
    const { ok, errorCount } = testParse(
      'S <- "x" / "y" / "z" ;',
      'z'
    );
    expect(ok).toBe(true);
    expect(errorCount).toBe(0);
  });

  test('FR04-with recovery', () => {
    const { ok, errorCount, skippedStrings } = testParse(
      'S <- "x"+ "!" / "fallback" ;',
      'xZx!'
    );
    expect(ok).toBe(true);
    expect(errorCount).toBe(1);
    expect(skippedStrings).toContain('Z');
  });

  test('FR05-fallback', () => {
    const { ok, errorCount } = testParse(
      'S <- "a" "b" / "x" ;',
      'x'
    );
    expect(ok).toBe(true);
    expect(errorCount).toBe(0);
  });

  test('FR06-none match', () => {
    const { ok } = testParse(
      'S <- "a" / "b" / "c" ;',
      'x'
    );
    expect(ok).toBe(false);
  });

  test('FR07-nested', () => {
    const { ok, errorCount } = testParse(
      'S <- ("a" / "b") / "c" ;',
      'b'
    );
    expect(ok).toBe(true);
    expect(errorCount).toBe(0);
  });

  test('FR08-deep nested', () => {
    const { ok, errorCount } = testParse(
      'S <- (("a")) ;',
      'a'
    );
    expect(ok).toBe(true);
    expect(errorCount).toBe(0);
  });
});

File: dart/test/parser/first_test.dart
// ===========================================================================
// SECTION 8: FIRST (ORDERED CHOICE) (15 tests)
// ===========================================================================

import 'package:test/test.dart';

import '../test_utils.dart';

void main() {
  test('FR01-match 1st', () {
    final (ok, err, _) = testParse(
      'S <- "abc" / "ab" / "a" ;',
      'abc',
    );
    expect(ok, isTrue, reason: 'should succeed');
    expect(err, equals(0), reason: 'should have 0 errors');
  });

  test('FR02-match 2nd', () {
    final (ok, err, _) = testParse(
      'S <- "xyz" / "abc" ;',
      'abc',
    );
    expect(ok, isTrue, reason: 'should succeed');
    expect(err, equals(0), reason: 'should have 0 errors');
  });

  test('FR03-match 3rd', () {
    final (ok, err, _) = testParse(
      'S <- "x" / "y" / "z" ;',
      'z',
    );
    expect(ok, isTrue, reason: 'should succeed');
    expect(err, equals(0), reason: 'should have 0 errors');
  });

  test('FR04-with recovery', () {
    final (ok, err, skip) = testParse(
      'S <- "x"+ "!" / "fallback" ;',
      'xZx!',
    );
    expect(ok, isTrue, reason: 'should succeed');
    expect(err, equals(1), reason: 'should have 1 error');
    expect(skip.contains('Z'), isTrue, reason: 'should skip Z');
  });

  test('FR05-fallback', () {
    final (ok, err, _) = testParse(
      'S <- "a" "b" / "x" ;',
      'x',
    );
    expect(ok, isTrue, reason: 'should succeed');
    expect(err, equals(0), reason: 'should have 0 errors');
  });

  test('FR06-none match', () {
    final (ok, _, _) = testParse(
      'S <- "a" / "b" / "c" ;',
      'x',
    );
    expect(ok, isFalse, reason: 'should fail');
  });

  test('FR07-nested', () {
    final (ok, err, _) = testParse(
      'S <- ("a" / "b") / "c" ;',
      'b',
    );
    expect(ok, isTrue, reason: 'should succeed');
    expect(err, equals(0), reason: 'should have 0 errors');
  });

  test('FR08-deep nested', () {
    final (ok, err, _) = testParse(
      'S <- (("a")) ;',
      'a',
    );
    expect(ok, isTrue, reason: 'should succeed');
    expect(err, equals(0), reason: 'should have 0 errors');
  });
}
